<a href="https://colab.research.google.com/github/RafaelRamos19/Corretor-ortogr-fico-com-word2vac/blob/main/Corretor.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
with open("artigos.txt", 'r') as f:
  artigos = f.read()
print(artigos[:100])




imagem 

Temos a seguinte classe que representa um usuário no nosso sistema:

java

Para salvar u


In [ ]:
import nltk
nltk.download('punkt')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

In [ ]:
palavras_separadas = nltk.tokenize.word_tokenize(artigos)

In [ ]:
palavras_separadas

['imagem',
 'Temos',
 'a',
 'seguinte',
 'classe',
 'que',
 'representa',
 'um',
 'usuário',
 'no',
 'nosso',
 'sistema',
 ':',
 'java',
 'Para',
 'salvar',
 'um',
 'novo',
 'usuário',
 ',',
 'várias',
 'validações',
 'são',
 'feitas',
 ',',
 'como',
 'por',
 'exemplo',
 ':',
 'Ver',
 'se',
 'o',
 'nome',
 'só',
 'contém',
 'letras',
 ',',
 '[',
 '*',
 '*',
 'o',
 'CPF',
 'só',
 'números',
 '*',
 '*',
 ']',
 'e',
 'ver',
 'se',
 'o',
 'usuário',
 'possui',
 'no',
 'mínimo',
 '18',
 'anos',
 '.',
 'Veja',
 'o',
 'método',
 'que',
 'faz',
 'essa',
 'validação',
 ':',
 'java',
 'Suponha',
 'agora',
 'que',
 'eu',
 'tenha',
 'outra',
 'classe',
 ',',
 'a',
 'classe',
 '`',
 'Produto',
 '`',
 ',',
 'que',
 'contém',
 'um',
 'atributo',
 'nome',
 'e',
 'eu',
 'quero',
 'fazer',
 'a',
 'mesma',
 'validação',
 'que',
 'fiz',
 'para',
 'o',
 'nome',
 'do',
 'usuário',
 ':',
 'Ver',
 'se',
 'só',
 'contém',
 'letras',
 '.',
 'E',
 'aí',
 '?',
 'Vou',
 'criar',
 'outro',
 'método',
 'para',
 'faz

In [ ]:
def separa_palavras(lista_tokens):
  lista_palavras = []
  for token in lista_tokens:
    if token.isalpha():
      lista_palavras.append(token)
  return lista_palavras



In [ ]:
lista_tokens = nltk.tokenize.word_tokenize(artigos)
lista_palavras = separa_palavras(lista_tokens)
print("O número de palavras é:",len(lista_palavras))

O número de palavras é: 403031


esse é o número total de palavras que aparecem no texto porém elas estão repetidas, dessa maneira deve-se normaliza-las

In [ ]:
def normalizacao(lista_palavras):
  lista_normalizada = []
  for palavra in lista_palavras:
    lista_normalizada.append(palavra.lower())
  return lista_normalizada

lista_normalizada = normalizacao(lista_palavras)
lista_normalizada

['imagem',
 'temos',
 'a',
 'seguinte',
 'classe',
 'que',
 'representa',
 'um',
 'usuário',
 'no',
 'nosso',
 'sistema',
 'java',
 'para',
 'salvar',
 'um',
 'novo',
 'usuário',
 'várias',
 'validações',
 'são',
 'feitas',
 'como',
 'por',
 'exemplo',
 'ver',
 'se',
 'o',
 'nome',
 'só',
 'contém',
 'letras',
 'o',
 'cpf',
 'só',
 'números',
 'e',
 'ver',
 'se',
 'o',
 'usuário',
 'possui',
 'no',
 'mínimo',
 'anos',
 'veja',
 'o',
 'método',
 'que',
 'faz',
 'essa',
 'validação',
 'java',
 'suponha',
 'agora',
 'que',
 'eu',
 'tenha',
 'outra',
 'classe',
 'a',
 'classe',
 'produto',
 'que',
 'contém',
 'um',
 'atributo',
 'nome',
 'e',
 'eu',
 'quero',
 'fazer',
 'a',
 'mesma',
 'validação',
 'que',
 'fiz',
 'para',
 'o',
 'nome',
 'do',
 'usuário',
 'ver',
 'se',
 'só',
 'contém',
 'letras',
 'e',
 'aí',
 'vou',
 'criar',
 'outro',
 'método',
 'para',
 'fazer',
 'a',
 'mesma',
 'validação',
 'ou',
 'criar',
 'uma',
 'interface',
 'ou',
 'uma',
 'classe',
 'que',
 'tanto',
 'usuario

In [ ]:
import string
pontuacao = string.punctuation

def retira_pontuacao(lista_palavras):
  lista_sem_pontuacao = []
  for palavra in lista_palavras:
    if palavra not in pontuacao:
      lista_sem_pontuacao.append(palavra)

  return lista_sem_pontuacao

lista_sem_pontuacao = retira_pontuacao(lista_palavras)
lista_sem_pontuacao

['imagem',
 'Temos',
 'a',
 'seguinte',
 'classe',
 'que',
 'representa',
 'um',
 'usuário',
 'no',
 'nosso',
 'sistema',
 'java',
 'Para',
 'salvar',
 'um',
 'novo',
 'usuário',
 'várias',
 'validações',
 'são',
 'feitas',
 'como',
 'por',
 'exemplo',
 'Ver',
 'se',
 'o',
 'nome',
 'só',
 'contém',
 'letras',
 'o',
 'CPF',
 'só',
 'números',
 'e',
 'ver',
 'se',
 'o',
 'usuário',
 'possui',
 'no',
 'mínimo',
 'anos',
 'Veja',
 'o',
 'método',
 'que',
 'faz',
 'essa',
 'validação',
 'java',
 'Suponha',
 'agora',
 'que',
 'eu',
 'tenha',
 'outra',
 'classe',
 'a',
 'classe',
 'Produto',
 'que',
 'contém',
 'um',
 'atributo',
 'nome',
 'e',
 'eu',
 'quero',
 'fazer',
 'a',
 'mesma',
 'validação',
 'que',
 'fiz',
 'para',
 'o',
 'nome',
 'do',
 'usuário',
 'Ver',
 'se',
 'só',
 'contém',
 'letras',
 'E',
 'aí',
 'Vou',
 'criar',
 'outro',
 'método',
 'para',
 'fazer',
 'a',
 'mesma',
 'validação',
 'Ou',
 'criar',
 'uma',
 'interface',
 'ou',
 'uma',
 'classe',
 'que',
 'tanto',
 'Usuario

In [ ]:
len(set(lista_sem_pontuacao)) #dessa maneira conseguimos tirar a repetição das palavras

21496

Agora conseguimos ver que o corretor com essa base de dados ele irá conseguir conrigir 21496 palavras

### Construção do modelo

O modelo irá funcionar recebendo uma palavra escrita de forma errada e irá retornar a palavra escrita de maneira certa

Sendo assim iremos inserir letas na palavra

Para que essa operação seja generalizada, iremos pegar cada posição da palavra em questão e inserir caracteres com o resto da palavra ao seu redor, até uma palavra existente ser formada. Isso pode cobrir um erro cometido em qualquer posição da palavra

In [ ]:
def insere_letras(fatias):
  novas_palavras = []
  letras = 'abcdefghijklmnopqrstuvwxyzàáâãèéêìíîòóôõùúûç'
  for E,D in fatias:
    for letra in letras:
      novas_palavras.append(E + letra + D)
  return novas_palavras

In [ ]:
palavra_exemplo = 'proocuar'
def gerador_palavras(palavra):
  fatias = []
  for i in range(len(palavra)+1):
    fatias.append((palavra[:i], palavra[i:]))
  palavras_geradas = insere_letras(fatias)
  return palavras_geradas


Para criar o gerador vamos ver qual é combinação de maior probabilidade de ser a palavra certa

Na função max podemos chamar uma outra função (vamos criar uma de probabilidade que irá retornar a probabilidade de cada variação de palavra, a max irá apenas tirar a maior delas)

In [ ]:
total_palavras = len(lista_sem_pontuacao)
frequencia = nltk.FreqDist(lista_sem_pontuacao) #irá registrar a frequência de cada palavra do artigo

In [ ]:
def probabilidade(palavra):
  return frequencia[palavra]/total_palavras

In [ ]:
def corretor(palavra):
  palavras_geradas = gerador_palavras(palavra)
  palavra_correta = max(palavras_geradas, key = probabilidade)
  return palavra_correta



In [ ]:
corretor('lgica')

'lógica'

In [ ]:
def cria_dados_teste(nome_do_arquivo):
  lista_de_palavras_teste = []
  f = open(nome_do_arquivo, 'r')
  for linha in f:
    correta, errada = linha.split()
    lista_de_palavras_teste.append((correta, errada))
  f.close()
  return lista_de_palavras_teste

lista_teste = cria_dados_teste('/content/palavras.txt')
lista_teste

[('podemos', 'pyodemos'),
 ('esse', 'esje'),
 ('já', 'jrá'),
 ('nosso', 'nossov'),
 ('são', 'sãêo'),
 ('dos', 'dosa'),
 ('muito', 'muifo'),
 ('imagem', 'iômagem'),
 ('sua', 'ósua'),
 ('também', 'tambéùm'),
 ('ele', 'eme'),
 ('fazer', 'èazer'),
 ('temos', 'temfs'),
 ('essa', 'eàssa'),
 ('quando', 'quaôdo'),
 ('vamos', 'vamvos'),
 ('sobre', 'hsobre'),
 ('java', 'sjava'),
 ('das', 'daõs'),
 ('agora', 'agorah'),
 ('está', 'eòtá'),
 ('cada', 'céda'),
 ('mesmo', 'zmesmo'),
 ('nos', 'noâ'),
 ('forma', 'fobma'),
 ('seja', 'sejéa'),
 ('então', 'enêão'),
 ('criar', 'èriar'),
 ('código', 'cóeigo'),
 ('caso', 'casío'),
 ('exemplo', 'áexemplo'),
 ('tem', 'tĩem'),
 ('usuário', 'usuárôio'),
 ('dados', 'dfados'),
 ('python', 'pgthon'),
 ('nossa', 'nossah'),
 ('além', 'alémè'),
 ('assim', 'asõim'),
 ('ter', 'teb'),
 ('até', 'atĩ'),
 ('bem', 'âem'),
 ('design', 'desigen'),
 ('trabalho', 'trabalàho'),
 ('foi', 'foo'),
 ('apenas', 'apenaũ'),
 ('empresa', 'empresà'),
 ('valor', 'valíor'),
 ('será', 'serr')

In [ ]:
def avaliador(testes):
  numero_de_palavras = len(testes)
  acertou = 0
  for certo, errado in testes:
    if corretor(errado) == certo:
      acertou +=1
  taxa_de_acerto = acertou/numero_de_palavras
  return print("A acurácia é de:", round(taxa_de_acerto,4)*100,'%')

avaliador(lista_teste)

A acurácia é de: 1.08 %


Nosso corretor inda não está tão bom pois ele corrige um erro bem específico

In [ ]:
def deletando_caracter(fatias):
  novas_palavras = []
  for E,D in fatias:
      novas_palavras.append(E + D[1:])
  return novas_palavras


In [ ]:
palavra_exemplo = 'proocuar'
def gerador_palavras(palavra):
  fatias = []
  for i in range(len(palavra)+1):
    fatias.append((palavra[:i], palavra[i:]))
  palavras_geradas = insere_letras(fatias)
  palavras_geradas += deletando_caracter(fatias)
  return palavras_geradas

gerador_palavras('lóigica')

['alóigica',
 'blóigica',
 'clóigica',
 'dlóigica',
 'elóigica',
 'flóigica',
 'glóigica',
 'hlóigica',
 'ilóigica',
 'jlóigica',
 'klóigica',
 'llóigica',
 'mlóigica',
 'nlóigica',
 'olóigica',
 'plóigica',
 'qlóigica',
 'rlóigica',
 'slóigica',
 'tlóigica',
 'ulóigica',
 'vlóigica',
 'wlóigica',
 'xlóigica',
 'ylóigica',
 'zlóigica',
 'àlóigica',
 'álóigica',
 'âlóigica',
 'ãlóigica',
 'èlóigica',
 'élóigica',
 'êlóigica',
 'ìlóigica',
 'ílóigica',
 'îlóigica',
 'òlóigica',
 'ólóigica',
 'ôlóigica',
 'õlóigica',
 'ùlóigica',
 'úlóigica',
 'ûlóigica',
 'çlóigica',
 'laóigica',
 'lbóigica',
 'lcóigica',
 'ldóigica',
 'leóigica',
 'lfóigica',
 'lgóigica',
 'lhóigica',
 'lióigica',
 'ljóigica',
 'lkóigica',
 'llóigica',
 'lmóigica',
 'lnóigica',
 'loóigica',
 'lpóigica',
 'lqóigica',
 'lróigica',
 'lsóigica',
 'ltóigica',
 'luóigica',
 'lvóigica',
 'lwóigica',
 'lxóigica',
 'lyóigica',
 'lzóigica',
 'làóigica',
 'láóigica',
 'lâóigica',
 'lãóigica',
 'lèóigica',
 'léóigica',
 'lêóigica',

In [ ]:
avaliador(lista_teste)

A acurácia é de: 41.4 %


Nós tivemos um upgrade muito grande em relação a como o programa estava desde a ultima implementação

In [ ]:
def troca_letras(fatias):
  novas_palavras = []
  letras = 'abcdefghijklmnopqrstuvwxyzàáâãèéêìíîòóôõùúûç'
  for E,D in fatias:
    for letra in letras:
      novas_palavras.append(E + letra + D[1:])
  return novas_palavras

def gerador_palavras(palavra):
  fatias = []
  for i in range(len(palavra)+1):
    fatias.append((palavra[:i], palavra[i:]))
  palavras_geradas = insere_letras(fatias)
  palavras_geradas += deletando_caracter(fatias)
  palavras_geradas += troca_letras(fatias)
  return palavras_geradas

gerador_palavras('ligica')

['aligica',
 'bligica',
 'cligica',
 'dligica',
 'eligica',
 'fligica',
 'gligica',
 'hligica',
 'iligica',
 'jligica',
 'kligica',
 'lligica',
 'mligica',
 'nligica',
 'oligica',
 'pligica',
 'qligica',
 'rligica',
 'sligica',
 'tligica',
 'uligica',
 'vligica',
 'wligica',
 'xligica',
 'yligica',
 'zligica',
 'àligica',
 'áligica',
 'âligica',
 'ãligica',
 'èligica',
 'éligica',
 'êligica',
 'ìligica',
 'íligica',
 'îligica',
 'òligica',
 'óligica',
 'ôligica',
 'õligica',
 'ùligica',
 'úligica',
 'ûligica',
 'çligica',
 'laigica',
 'lbigica',
 'lcigica',
 'ldigica',
 'leigica',
 'lfigica',
 'lgigica',
 'lhigica',
 'liigica',
 'ljigica',
 'lkigica',
 'lligica',
 'lmigica',
 'lnigica',
 'loigica',
 'lpigica',
 'lqigica',
 'lrigica',
 'lsigica',
 'ltigica',
 'luigica',
 'lvigica',
 'lwigica',
 'lxigica',
 'lyigica',
 'lzigica',
 'làigica',
 'láigica',
 'lâigica',
 'lãigica',
 'lèigica',
 'léigica',
 'lêigica',
 'lìigica',
 'líigica',
 'lîigica',
 'lòigica',
 'lóigica',
 'lôigica',
 'lõ

In [ ]:
avaliador(lista_teste)

A acurácia é de: 76.88000000000001 %


A nossa acurácia melhorou muito 

Vamos agora inverter uma letra pela outra

In [ ]:
def inverte_letras(fatias):
  novas_palavras = []
  for E,D in fatias:
    if len(D) >1:
      novas_palavras.append(E + D[1] + D[0] + D[2:])
  return novas_palavras

def gerador_palavras(palavra):
  fatias = []
  for i in range(len(palavra)+1):
    fatias.append((palavra[:i], palavra[i:]))
  palavras_geradas = insere_letras(fatias)
  palavras_geradas += deletando_caracter(fatias)
  palavras_geradas += troca_letras(fatias)
  palavras_geradas +=inverte_letras(fatias)
  return palavras_geradas

gerador_palavras('lgoica')

['algoica',
 'blgoica',
 'clgoica',
 'dlgoica',
 'elgoica',
 'flgoica',
 'glgoica',
 'hlgoica',
 'ilgoica',
 'jlgoica',
 'klgoica',
 'llgoica',
 'mlgoica',
 'nlgoica',
 'olgoica',
 'plgoica',
 'qlgoica',
 'rlgoica',
 'slgoica',
 'tlgoica',
 'ulgoica',
 'vlgoica',
 'wlgoica',
 'xlgoica',
 'ylgoica',
 'zlgoica',
 'àlgoica',
 'álgoica',
 'âlgoica',
 'ãlgoica',
 'èlgoica',
 'élgoica',
 'êlgoica',
 'ìlgoica',
 'ílgoica',
 'îlgoica',
 'òlgoica',
 'ólgoica',
 'ôlgoica',
 'õlgoica',
 'ùlgoica',
 'úlgoica',
 'ûlgoica',
 'çlgoica',
 'lagoica',
 'lbgoica',
 'lcgoica',
 'ldgoica',
 'legoica',
 'lfgoica',
 'lggoica',
 'lhgoica',
 'ligoica',
 'ljgoica',
 'lkgoica',
 'llgoica',
 'lmgoica',
 'lngoica',
 'logoica',
 'lpgoica',
 'lqgoica',
 'lrgoica',
 'lsgoica',
 'ltgoica',
 'lugoica',
 'lvgoica',
 'lwgoica',
 'lxgoica',
 'lygoica',
 'lzgoica',
 'làgoica',
 'lágoica',
 'lâgoica',
 'lãgoica',
 'lègoica',
 'légoica',
 'lêgoica',
 'lìgoica',
 'lígoica',
 'lîgoica',
 'lògoica',
 'lógoica',
 'lôgoica',
 'lõ

In [ ]:
avaliador(lista_teste)

A acurácia é de: 76.88000000000001 %


Possuimos uma taxa de acerto, porém não está claro nela quanto daquelas palavras são desconhecidas para nosso algorítmo

In [ ]:
def avaliador(testes,vocabulario):
  numero_de_palavras = len(testes)
  acertou = 0
  desconhecida = 0
  for certo, errado in testes:
    if corretor(errado) == certo:
      acertou +=1
    else:
      desconhecida += (certo not in vocabulario)
  taxa_de_acerto = round(acertou*100/numero_de_palavras,2)
  taxa_desconhecida = round(desconhecida*100/numero_de_palavras,2)
  return print(f"A acurácia é de: {taxa_de_acerto}%, palavra desconhecida é {taxa_desconhecida}%")


vocabulario = set(lista_normalizada)
avaliador(lista_teste,vocabulario)

A acurácia é de: 76.88%, palavra desconhecida é 6.99%


Agora foi possível melhorar o corretor, entretanto não é possível corrigir palavras como: "loiigica", para isso, teremos que abrangir ainda mais nosso código
O gerador turbinado irá receber palavras já geradas nas funções anteriores

In [ ]:
def gerador_turbinado(palavras_geradas):
  novas_palavras = []
  for palavra in palavras_geradas:
    novas_palavras += gerador_palavras(palavra)
  return novas_palavras

palavras_g = gerador_turbinado(gerador_palavras("lóiigica"))
"lógica" in palavras_g

True

In [ ]:
len(palavras_g)

691744

Ou seja, temos 691744 variações da mesma palavra
precisamos de alguma maneira eliminar as palavras que não são candidatas a serem certas

Para resolver o problema, vamos modificar a função corretor, ela irá retornar uma nova lista apenas com a palavra a ser corrigida e as palavras existentes no vocabulário

In [ ]:
def novo_corretor(palavra):
  palavras_geradas = gerador_palavras(palavra)
  palavras_geradas_turbinado = gerador_turbinado(palavras_geradas)
  todas_palavras = set(palavras_geradas + palavras_geradas_turbinado)
  candidatos = [palavra]
  for palavra in todas_palavras:
    if palavra in vocabulario:
      candidatos.append(palavra)
  palavra_correta = max(candidatos, key = probabilidade)
  return palavra_correta

novo_corretor("lóiigica")

'lógica'

Observe que as únicas palavras relevantes para o corretor foram apenas duas palavras

In [ ]:
def avaliador(testes,vocabulario):
  numero_de_palavras = len(testes)
  acertou = 0
  desconhecida = 0
  for certo, errado in testes:
    desconhecida += (certo not in vocabulario)
    if novo_corretor(errado) == certo:
      acertou +=1
    else:
      print(errado + "-" + corretor(errado) + '-'+novo_corretor(errado) )
  taxa_de_acerto = round(acertou*100/numero_de_palavras,2)
  taxa_desconhecida = round(desconhecida*100/numero_de_palavras,2)
  return print(f"A acurácia é de: {taxa_de_acerto}%, palavra desconhecida é {taxa_desconhecida}%")


vocabulario = set(lista_normalizada)
avaliador(lista_teste,vocabulario)

esje-esse-se
sãêo-são-não
dosa-dos-do
eme-em-de
eàssa-essa-esse
daõs-das-da
céda-cada-da
noâ-no-o
enêão-então-não
tĩem-tem-em
nossah-nossa-nosso
teb-tem-de
atĩ-até-a
âem-em-de
foo-foi-o
serr-ser-se
van-vai-a
çeus-seus-seu
eû-e-de
temeo-tempo-temos
semre-sempre-ser
elaá-ela-ele
síó-só-se
siàe-site-se
seém-sem-em
peln-pelo-ele
aléra-alura-agora
tdia-dia-da
tuúo-tudo-tipo
jé-é-de
sãô-são-não
odos-dos-do
siua-sua-seu
elpe-ele-esse
teos-temos-os
eũsa-essa-esse
vjmos-vamos-temos
dms-dos-de
cava-cada-para
ános-nos-os
èaso-caso-as
túem-tem-em
daáos-dados-dos
nossk-nosso-nosso
tãer-ter-ser
vté-até-é
búm-bem-um
sçerá-será-ser
uai-vai-a
sâus-seus-seu
ìeu-seu-de
fual-qual-sua
elal-ela-ele
skó-só-se
secm-sem-em
aluéa-alura-além
dil-dia-de
sód-só-se
eúaa-aeúaa-essa
ró-só-de
dĩaz-adĩaz-da
correptor-corretor-correto
trtica-tática-prática
ewpoderamento-aewpoderamento-ewpoderamento
lifux-linux-flux
îgato-gato-fato
cakvalo-acakvalo-cakvalo
canelac-acanelac-janela
tênisy-atênisy-tênisy
anciosa-aanciosa-an

Observe que em muitos casos o corretor conseguiu resolver o problema e o novo corretor acabou errando

Desse modo, é mais vantajoso continuar com o antigo corretor

In [ ]:
def avaliador(testes,vocabulario):
  numero_de_palavras = len(testes)
  acertou = 0
  desconhecida = 0
  for certo, errado in testes:
    desconhecida += (certo not in vocabulario)
    if corretor(errado) == certo:
      acertou +=1
  taxa_de_acerto = round(acertou*100/numero_de_palavras,2)
  taxa_desconhecida = round(desconhecida*100/numero_de_palavras,2)
  return print(f"A acurácia é de: {taxa_de_acerto}%, palavra desconhecida é {taxa_desconhecida}%")


vocabulario = set(lista_normalizada)
avaliador(lista_teste,vocabulario)

A acurácia é de: 76.88%, palavra desconhecida é 6.99%


Observe para o exemplo da palavra lógica para os dois corretores

In [ ]:
palavra ='lgica'
print("Correção com o antigo corretor:", corretor(palavra))
print("Correção com o novo corretor:", novo_corretor(palavra))

Correção com o antigo corretor: lógica
Correção com o novo corretor: fica
